proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,692 @@
+  variants would match the value "bar".
+- There must be a variant of the resource for every value of a key that is
+  going to be present.  For example, if clients will send constraints on the
+  `env` key requiring the value to be one of `prod`, `test`, or `qa`, then
+  you must have each of those three variants of the resource.  (Failure
+  to do this will result in the server acting as if the requested
+  resource does not exist.)
+
+#### Transport Protocol Changes
+
+The following message will be added to represent a subscription to a
+resource by name with associated dynamic parameters:
+
+```proto
+// A specification of a resource used when subscribing or unsubscribing.
+message ResourceLocator {
+  // The resource name to subscribe to.
+  string name = 1;
+
+  // A set of dynamic parameters used to match against the dynamic parameter
+  // constraints on the resource. This allows clients to select between
+  // multiple variants of the same resource.
+  map<string, string> dynamic_parameters = 2;
+}
+```
+
+The following new field will be added to `DiscoveryRequest`, to allow clients
+to specify dynamic parameters when subscribing to a resource:
+
+```proto
+  // Alternative to resource_names field that allows specifying cache
+  // keys along with each resource name. Clients that populate this field
+  // must be able to handle responses from the server where resources are
+  // wrapped in a Resource message.
+  repeated ResourceLocator resource_locators = 7;
+```
+
+Similarly, the following fields will be added to `DeltaDiscoveryRequest`:
+
+```proto
+  // Alternative to resource_names_subscribe field that allows specifying cache
+  // keys along with each resource name.
+  repeated ResourceLocator resource_locators_subscribe = 8;
+
+  // Alternative to resource_names_unsubscribe field that allows specifying cache
+  // keys along with each resource name.
+  repeated ResourceLocator resource_locators_unsubscribe = 9;
+```
+
+The following field will be added to the `Resource` message, to allow the
+server to return the dynamic parameters associated with each resource:
+
+```proto
+  // Dynamic parameter constraints associated with this resource. To be used
+  // by client-side caches (including xDS proxies) when matching subscribed
+  // resource locators.
+  DynamicParameterConstraints dynamic_parameter_constraints = 8;
+```
+
+### Migrating From Node Metadata
+
+Today, the equivalent of dynamic parameter constraints is node metadata,
+which can be used by servers to determine the set of resources to send
+for LDS and CDS wildcard subscriptions or to determine the contents of
+other resources (e.g., to select individual routes to be included in an
+RDS resource).  For transition purposes, this mechanism can continue
+to be supported by the client performing direct translation of node
+metadata to exact-match constraints.  For example, if the node metadata
+contains the entry `env=prod`, this would be translated to a constraint
```

Comment:
This seems broken. I expected `env=prod` to be translated to dynamic parameters `{env=prod}` on client-side if we're talking about doing translation. But this is talking about _constraints_ which means it is server-side. It seems clients/proxies wouldn't be able to match these constraints in response objects, because their dynamic properties _don't_ include `env`.